### PR TITLE
Add target settings attribute to toolchain rule

### PR DESCRIPTION
**Pull request by comius**
This is the implementation of proposal https://docs.google.com/document/d/13yXhOQeMZQmlW3obyZ7fJP3sm3eVLbK3Nuk2LZjcZiU/edit.

The added `targetSetting` attribute, can reference `config_setting` rules. The attribute is resolved in target configuration and the returned `ConfigMatchingProvider` already has the value if this toolchain matches.

Difference to the proposal is that `targetSettings` are a `LABEL_LIST` where all the settings need to be satisfied. This might increase a bit usability. It makes the implementation a bit nicer, because instead of special handling when the value is not present, a list can uniformly handled.

### CODE REVIEW

**Comment by comius** [comment]
cc @lberki, @oquenchil, @c-mita

File: src/main/java/com/google/devtools/build/lib/skyframe/SingleToolchainResolutionFunction.java [hunk 1]
```diff
@@ -139,7 +140,8 @@ private static SingleToolchainResolutionValue resolveConstraints(
       }
 
       // Make sure the target platform matches.
-      if (!checkConstraints(
+      if (!toolchain.targetSettings().stream().allMatch(ConfigMatchingProvider::matches)
```
Comment (katre) [inline src/main/java/com/google/devtools/build/lib/skyframe/SingleToolchainResolutionFunction.java:143]: We should probably enhance the output of --toolchain_resolution_debug to clarify when a toolchain is rejected for settings instead of target/exec constraints.

File: src/main/java/com/google/devtools/build/lib/rules/platform/ToolchainRule.java [hunk 2]
```diff
@@ -78,6 +81,14 @@ public RuleClass build(RuleClass.Builder builder, RuleDefinitionEnvironment env)
                 .mandatoryProviders(ConstraintValueInfo.PROVIDER.id())
                 .allowedFileTypes()
                 .nonconfigurable("part of toolchain configuration"))
+        /* <!-- #BLAZE_RULE(toolchain).ATTRIBUTE(target_settings) -->
+        A list of <code>config_setting</code>s that must be satisfied by the target
+        build configuration in order for this toolchain to be selected for a target building.
```
Comment (katre) [inline src/main/java/com/google/devtools/build/lib/rules/platform/ToolchainRule.java:86]: How about this wording?

```
A list of <code>config_setting</code>s that must be satisfied by the target configuration in order for this toolchain to be selected during toolchain resolution.
```

**Review by katre — CHANGES_REQUESTED** [review]
Is the design review for this finished? I think we've discussed my concerns (and I see you're only implementing target_setting, not exec_setting), but has @cushon reviewed the design yet?

File: src/main/java/com/google/devtools/build/lib/skyframe/SingleToolchainResolutionFunction.java (hunk 1 above)
Comment (comius) [reply src/main/java/com/google/devtools/build/lib/skyframe/SingleToolchainResolutionFunction.java]: Done.

File: src/main/java/com/google/devtools/build/lib/rules/platform/ToolchainRule.java (hunk 2 above)
Comment (comius) [reply src/main/java/com/google/devtools/build/lib/rules/platform/ToolchainRule.java]: Done.

**Comment by comius** [comment]
> Is the design review for this finished? I think we've discussed my concerns (and I see you're only implementing target_setting, not exec_setting), but has @cushon reviewed the design yet?

No, still waiting for @cushon. Don't want to miss Bazel release though, because this could prolong everything for another month....

**Review by katre — APPROVED** [review]
LGTM, acknowledging that we may need to change details after the review finishes.
